### ISSUE TITLE
cubemap tricks

### ISSUE THREAD

**Issue by AugmentedRealityCat** [GeorgLegato/sd-webui-panorama-viewer#23]
Here is an example of a workflow to create seamless cubemaps from an equirectangular panorama. I am convinced this could be automated.

1. Create or import an **equirectangular panorama**.

![00054](https://user-images.githubusercontent.com/114955776/227856271-bd9dcb13-7813-46d1-a151-879ac0cb686b.png)

2. Load it in a viewer, set the camera **FOV to 90 degrees**.
3. **Tilt down the camera** until it points directly at the nadir (bottom).
4. **Render** that view as a square image.

![nadir_view_from_spherical_projection](https://user-images.githubusercontent.com/114955776/227856381-1350cfa2-e6eb-49e4-a93c-f5a9e76f2846.png)

5. Load that image as a source for **IMG2IMG inpainting with mask**.
6. Apply a **circular mask** - a good recipe is one with a gradient going from 25% to 50% in diameter. 

![circular_mask_25-50percentFeather](https://user-images.githubusercontent.com/114955776/227856644-d12089f1-b4fe-4d34-b720-856ca1f2e610.png)

7. You can set the mask blur to 0 in the A1111 WebUI if your mask is already feathered like mine.
8. Select an **inpainting model**, use **"background" as a prompt**, set **denoising to 0.8** and the **CFG to a low value like 3**.

![Screenshot 2023-03-27 at 01-10-00 Stable Diffusion copy_down](https://user-images.githubusercontent.com/114955776/227856452-a56ed7d5-49fd-4e09-a727-d4a8c160c01f.png)

9. Generate a picture. For this example I used the **Euler sampler with 50 steps**, which works well, but maybe we can optimize this.

![00025-2119308706](https://user-images.githubusercontent.com/114955776/227856599-8ae4e99b-8b56-43f3-9e2f-00656dbd8a10.png)

10. Since the content you generated is just inside the circular mask, the content all around is still **tiling seamlessly** with your panorama.
11. You **do the same for the zenith by pointing your camera up** and rendering another square

![Screenshot 2023-03-27 at 01-10-00 Stable Diffusion copy](https://user-images.githubusercontent.com/114955776/227856733-6d889caf-0cbe-4ccc-a05e-36928d28d247.png)

12. You **assemble** the 4 squares forming the 4 sides of our cube with the 2 new views we rendered for the top and bottom to create **the cubemap texture**. There are different configurations to arrange those 6 squares, but the **cross configuration shown below is the most popular**. If your viewer uses a different configuration, like having all the 6 squares in a single horizontal row, then just use that, it will work just as well, as long as the squares are in the right order and the proper orientation. 

![assemblage_en_croix copy](https://user-images.githubusercontent.com/114955776/227856863-c4a74ece-024a-4227-b7fe-5f277dbd91d5.png)

BONUS: if you convert that cubemap back into an equirectangular spherical map most of the time **it will fix polar distortion** and it will not show any pinching-of-the-UVs around the poles anymore ! You can test it with the example below and compare with the original spherical map at the top. 

![cube_to_sphere_8bit](https://user-images.githubusercontent.com/114955776/227858477-46c65afa-9474-40d1-aba9-1c89fd2d4402.png)

**Comment by GeorgLegato** [comment]
That is an incredible workflow, you are a magician.
thanks. I would like to put it to an extra document like "workflow-cube.md" into repository.

**Comment by GeorgLegato** [comment]
we could automate it, right? so the viewer becomes a toolkit step by step

**Comment by AugmentedRealityCat** [comment]
> That is an incredible workflow, you are a magician. thanks. I would like to put it to an extra document like "workflow-cube.md" into repository.

Go ahead, my goal is to help you and our community of users !
I did not check if you were using the cross-configuration for the cubemap though. If you use a different config, I can adjust my little workflow accordingly.

**Comment by AugmentedRealityCat** [comment]
> we could automate it, right? so the viewer becomes a toolkit step by step

Yes, that is my intention. At first it could be step-by-step much like the current EBSynth extension. But ultimately I think this can be fully automated.

I'm working on a similar manual workflow to create 360 HDRI images by stacking pictures simulating bracketed exposure and I also believe it could be automated at some point. I wish I had more time to work on it, but I am extremely busy at the moment !

**Comment by GeorgLegato** [comment]
the photoshpere viewer (with three.js inside) accepts that layout of a horizontal T

```
     top
left mid right rear
     bot
```
and list of tiles [1,2,3,4,5,6]

**Comment by AugmentedRealityCat** [comment]
New horizontal T layout version below:
![horizontal_T_cubemap](https://user-images.githubusercontent.com/114955776/228402988-f24962c4-01e5-4cb4-8365-198cc7eeb338.png)

**Comment by GeorgLegato** [comment]
ah nice.. if you want, you can try my branch, first draft implmentaton of Convert To Cubemap (available in txt2img yet)

**Comment by GeorgLegato** [comment]
Create an EQUI-IMG
Click on Pano Convert
Click on SendToOutpaint

![image](https://user-images.githubusercontent.com/7210708/228403846-630d3187-0dd0-4aed-af6d-dd13c4667ed6.png)
![image](https://user-images.githubusercontent.com/7210708/228403893-c489800e-1b1e-4b06-ac12-62226e36dcf2.png)
![image](https://user-images.githubusercontent.com/7210708/228404302-c35de6ad-e24e-4429-970a-fcf59b1a1137.png)

**Comment by AugmentedRealityCat** [comment]
I'll test that tomorrow when I get a minute - that looks like it will be a very useful function to fix both polar pinching and tiling on X axis.
